### PR TITLE
Implement Android Executors for Storage

### PR DESCRIPTION
**Pull request by maneesht**
[go/firebase-android-executors](http://goto.google.com/firebase-android-executors)

### CODE REVIEW

**Comment by maneesht** [comment]
@rlazo - It seems like creating a `LimitedConcurrencyExecutor` with only `1` as the concurrency causes many of the tests to fail, as for some reason a different thread calls the callback than the original one created. I fixed this by using a `SequentialExecutor`

File: firebase-storage/src/main/java/com/google/firebase/storage/StorageRegistrar.java [hunk 1]
```diff
@@ -17,35 +17,41 @@
...
 import java.util.Arrays;
 import java.util.List;
+import java.util.concurrent.Executor;
 
 /** @hide */
 @Keep
 @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
 public class StorageRegistrar implements ComponentRegistrar {
   private static final String LIBRARY_NAME = "fire-gcs";
+  Qualified<Executor> backgroundExecutor = Qualified.qualified(Blocking.class, Executor.class);
```
Comment (rlazo) [inline firebase-storage/src/main/java/com/google/firebase/storage/StorageRegistrar.java:37]: It's better to be consistent with naming, rename this to `blockingExecutor`

File: firebase-storage/src/main/java/com/google/firebase/storage/FirebaseStorageComponent.java [hunk 2]
```diff
@@ -35,10 +36,12 @@ class FirebaseStorageComponent {
   FirebaseStorageComponent(
       @NonNull FirebaseApp app,
       @Nullable Provider<InternalAuthProvider> authProvider,
-      @Nullable Provider<InternalAppCheckTokenProvider> appCheckProvider) {
+      @Nullable Provider<InternalAppCheckTokenProvider> appCheckProvider,
+      @NonNull Executor executor) {
```
Comment (rlazo) [inline firebase-storage/src/main/java/com/google/firebase/storage/FirebaseStorageComponent.java:40]: Annotate with the type of executor for clarity

File: firebase-storage/src/main/java/com/google/firebase/storage/StorageRegistrar.java (hunk 1 above)
Comment (maneesht) [reply firebase-storage/src/main/java/com/google/firebase/storage/StorageRegistrar.java]: Nice catch. Fixed

**Comment by maneesht** [comment]
> > @rlazo - It seems like creating a `LimitedConcurrencyExecutor` with only `1` as the concurrency causes many of the tests to fail, as for some reason a different thread calls the callback than the original one created. I fixed this by using a `SequentialExecutor`
> 
> I didn't see any changes to the tests , where did you made the change?

I was referring to the change made here: https://github.com/firebase/firebase-android-sdk/pull/4830/files#diff-8a3cd8845e80a432177d61c12b761f03d6da86be0b22b62ee0de1d5341a8fbefR44

**Review by rlazo — APPROVED** [review]

**Comment by rlazo** [comment]
If you sync past https://github.com/firebase/firebase-android-sdk/pull/4854 tests should start passing. Thanks!

File: firebase-storage/src/testUtil/java/com/google/firebase/storage/TestUploadHelper.java [hunk 3]
```diff
@@ -684,6 +684,7 @@ public static void cancelInProgressTask() {
 
   private static void verifyTaskCount(StorageReference reference, int expectedTasks) {
     List<UploadTask> globalUploadTasks = reference.getActiveUploadTasks();
+    System.out.println("Size: " + globalUploadTasks.size());
```
Comment (rlazo) [inline firebase-storage/src/testUtil/java/com/google/firebase/storage/TestUploadHelper.java:687]: Is this println inteded?

**Review by tonyjhuang — APPROVED** [review]

**Pull request merged**